tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+  unsigned char *buffer = calloc(size, sizeof(int));
+  unsigned char *index = buffer;
+
+  for (size_t i = 0; i < size; i++)
+  {
+    *index = (unsigned char) rand();
+    index += sizeof(int);
+  }
+
+  *nonce = buffer;
+  return 0;
+}
+
+//
+// negotiate_client_session_key()
+//
+int negotiate_client_session_key(int socket_fd,
+                                 EVP_PKEY_CTX * public_key_ctx,
+                                 EVP_PKEY_CTX * private_key_ctx,
+                                 unsigned char *id, size_t id_len,
+                                 unsigned char *expected_id,
+                                 size_t expected_id_len,
+                                 unsigned char **session_key,
+                                 size_t *session_key_len)
+{
+  // Generate nonce A
+  unsigned char *nonce_a = NULL;
+  size_t nonce_a_len = 0;
+
+  int result = generate_nonce(32, &nonce_a, &nonce_a_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to generate a nonce.");
+    return 1;
+  }
+
+  // Conduct NSL to obtain nonce B
+  unsigned char *nonce_b = NULL;
+  size_t nonce_b_len = 0;
+
+  unsigned char *request = NULL;
+  size_t request_len = 0;
+
+  unsigned char *response = calloc(8192, sizeof(unsigned char));
+  size_t response_len = 8192;
+
+  kmyth_log(LOG_INFO, "Sending nonce A: %zd bytes", nonce_a_len);
+
+  result = build_nonce_request(public_key_ctx,
+                               nonce_a, nonce_a_len,
+                               id, id_len, &request, &request_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to build the nonce request.");
+    return 1;
+  }
+
+  if (write(socket_fd, request, request_len) != request_len)
+  {
+    kmyth_log(LOG_ERR, "Failed to fully send nonce request message.");
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Successfully sent nonce A.");
+
+  ssize_t read_result = read(socket_fd, response, response_len);
+
+  if (-1 == read_result)
+  {
+    kmyth_log(LOG_ERR, "Failed to read the nonce response message.");
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Received %zd bytes", read_result);
+
+  free(request);
+  request = NULL;
+  request_len = 0;
+
+  unsigned char *received_nonce_a = NULL;
+  size_t received_nonce_a_len = 0;
+
+  unsigned char *received_id = NULL;
+  size_t received_id_len = 0;
+
+  result = parse_nonce_response(private_key_ctx,
+                                response, read_result,
+                                &received_nonce_a, &received_nonce_a_len,
+                                &nonce_b, &nonce_b_len,
+                                &received_id, &received_id_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to parse the nonce response.");
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Received nonce A: %zd bytes", received_nonce_a_len);
+  kmyth_log(LOG_INFO, "Received nonce B: %zd bytes", nonce_b_len);
+  kmyth_log(LOG_INFO, "Received ID: %.*s", received_id_len, received_id);
+
+  if (nonce_a_len != received_nonce_a_len)
```

Comment:
We should also check nonce_b_len to make sure that the nonce we received from the server matches what we expect (e.g. 32 bytes).

Reply:
Agreed. For now I'll check that the received nonces are of equal length.